File: src/day4/part2.py
import re

file = open("input", 'r')
lines = file.readlines()

passports = [passport.replace("\n", " ") for passport in ''.join(lines).split("\n\n")]

validentries = ['byr:', 'iyr:', 'eyr:', 'hgt:', 'hcl:', 'ecl:', 'pid:'] #cid:

def passportvalid(passport):
    entries=re.findall('...:', passport)
    return all([x in entries for x in validentries])

def deepvalidation(passport):
    if passportvalid(passport):
        keys =[s.replace(":", "") for s in re.findall('...:', passport)]
        vals =[s.replace(":", "") for s in re.findall(':[^ ]*', passport)]
        passdict = dict(zip(keys, vals))
        heightCheck = False
        if "in" in passdict['hgt']:
            heightCheck = (int(passdict['hgt'].replace("in","")) >= 59 and 
                    int(passdict['hgt'].replace("in", "")) <= 76)
        elif "cm" in passdict['hgt']:
            heightCheck= (int(passdict['hgt'].replace("cm","")) >= 150 and 
                    int(passdict['hgt'].replace("cm", "")) <= 193)
        else:
            False
        return all([
            int(passdict['byr']) >= 1920 and int(passdict['byr']) <= 2002,
            int(passdict['iyr']) >= 2010 and int(passdict['iyr']) <= 2020,
            int(passdict['eyr']) >= 2020 and int(passdict['eyr']) <= 2030,
            re.search('#[0-9,a-f]{6}', passdict['hcl']) != None and len(passdict['hcl']) == 7,
            heightCheck,
            [x == passdict['ecl'] for x in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']].count(True) == 1,
            re.search('^[0-9]{9}$', passdict['pid']) != None
            ])
    else:
        return False


print([deepvalidation(passport) for passport in passports].count(True))

File: src/day4/haskell/part2.hs
import Data.List.Split
import Text.Regex
import Data.Text (isInfixOf, pack)
main = do
  input <- readFile "input"
  let l =  map (map (\x -> if x /= '\n' then x else ' ')) $ splitOn "\n\n" input
  let valids = filter (isValid) l
  let parsed = map (map (splitOn ":") ) $ map (splitOn " ") valids
  putStrLn $ show $ length $ filter (deepValid) parsed

isValid s = all (==True) [x `isInfixOf` (pack s) | x <- validentries]
  where validentries = map pack ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

deepValid :: [[String]] -> Bool
deepValid s = all (==True) $ map check s
  where
    check :: [String] -> Bool
    check [x, v] = case x of
                "byr" -> between 1920 2002 $ toInt v
                "iyr" -> between 2010 2020 $ toInt v
                "eyr" -> between 2020 2030 $ toInt v
                "hgt" -> checkHeight v
                "hcl" -> matchRegex (mkRegex "^#[0-9,a-z]{6}$") v /= Nothing
                "ecl" -> v `elem` ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
                "pid" -> matchRegex (mkRegex "^[0-9]{9}$") v /= Nothing
                _ -> True
    toInt x = read x :: Int
    between x y v = v <= y && v >= x
    checkHeight v = if ((pack "in")`isInfixOf`(pack v)) 
                       then between 59 76 $ getHeight v
                    else if ((pack "cm") `isInfixOf` (pack v))
                       then between 150 193 $ getHeight v
                    else False

    getHeight x = read (subRegex (mkRegex "in|cm") x "") :: Int

